c_src/ex_dtls/native.c

```diff
@@ -119,22 +144,44 @@ UNIFEX_TERM do_handshake(UnifexEnv *env, State *state, UnifexPayload *payload) {
       */
       DEBUG("Handshake finished successfully");
 
-      unsigned char *material = export_keying_material(state->ssl);
-      if (material == NULL) {
+      KeyingMaterial *keying_material = export_keying_material(state->ssl);
+      if (keying_material == NULL) {
         DEBUG("Cannot export keying material");
         exit(EXIT_FAILURE);
       }
 
-      DEBUG("Keying material %s", material);
+      int len = keying_material->len;
+      UnifexPayload *client_keying_material =
+          (UnifexPayload *)unifex_payload_alloc(env, UNIFEX_PAYLOAD_BINARY, len);
+      memcpy(client_keying_material->data, keying_material->client, len);
+      client_keying_material->size = len;
+
+      UnifexPayload *server_keying_material =
+          (UnifexPayload *)unifex_payload_alloc(env, UNIFEX_PAYLOAD_BINARY, len);
+      memcpy(server_keying_material->data, keying_material->server, len);
+      server_keying_material->size = len;
+
       if (dyn_buff->data_size == 0) {
+        UNIFEX_TERM res_term = do_handshake_result_finished(
+            env, state, client_keying_material, server_keying_material,
+            keying_material->protection_profile);
         dyn_buff_free(dyn_buff);
-        return do_handshake_result_finished(env, state, (char *)material);
+        unifex_payload_release(client_keying_material);
+        unifex_payload_release(server_keying_material);
+        return res_term;
       } else {
-        UnifexPayload *payload = (UnifexPayload *)unifex_payload_alloc(env, UNIFEX_PAYLOAD_BINARY, dyn_buff->data_size);
+        UnifexPayload *payload = (UnifexPayload *)unifex_payload_alloc(
+            env, UNIFEX_PAYLOAD_BINARY, dyn_buff->data_size);
         memcpy(payload->data, dyn_buff->data, dyn_buff->data_size);
         payload->size = (unsigned int)dyn_buff->data_size;
+        UNIFEX_TERM res_term = do_handshake_result_finished_with_packets(
+            env, state, client_keying_material, server_keying_material,
+            keying_material->protection_profile, payload);
         dyn_buff_free(dyn_buff);
-        return do_handshake_result_finished_with_packets(env, state, (char *)material, payload);
+        unifex_payload_release(payload);
+        unifex_payload_release(client_keying_material);
+        unifex_payload_release(server_keying_material);
```

Comment:
rather do the entire cleanup in a single place

Reply:
In different places I am cleaning up different things so I am not sure if this is a good idea?